File: side-projects/peri-sync/server/src/canonical.test.ts
// canonical.ts 单元测试：transcript 固定向量与 Crockford 归一化。
// 固定向量与 Rust `peri-tui/src/sync/canonical_test.rs` 一致。

import { describe, expect, it } from "bun:test";
import {
	PROTOCOL_TAG,
	b64urlDecode,
	b64urlEncode,
	normalizeSyncCode,
	timestampWithinSkew,
	transcript,
	utf8Bytes,
} from "./canonical";

describe("transcript", () => {
	it("复刻 Rust 固定向量（字段顺序、分隔符与时间戳逐字节稳定）", () => {
		expect(transcript("create", ["ch_abc", "dev_xyz"], 1715000000)).toBe(
			"peri-sync/v1|create|ch_abc|dev_xyz|1715000000",
		);
		expect(transcript("join", ["code_1", "ch_abc"], 0)).toBe(
			"peri-sync/v1|join|code_1|ch_abc|0",
		);
		expect(transcript("revoke", [], 42)).toBe("peri-sync/v1|revoke|42");
		expect(transcript("payload", ["c1", "3", "HASH"], 0)).toBe(
			"peri-sync/v1|payload|c1|3|HASH|0",
		);
	});

	it("PROTOCOL_TAG 固定", () => {
		expect(PROTOCOL_TAG).toBe("peri-sync/v1");
	});

	it("拒绝空 op / 空字段 / 含 | 片段", () => {
		expect(() => transcript("", ["a"], 1)).toThrow();
		expect(() => transcript("op|evil", ["a"], 1)).toThrow();
		expect(() => transcript("op", [""], 1)).toThrow();
		expect(() => transcript("op", ["a|b"], 1)).toThrow();
	});

	it("时间戳偏差判定（±300s 窗口）", () => {
		expect(timestampWithinSkew(1700, 2000, 300)).toBe(true);
		expect(timestampWithinSkew(2300, 2000, 300)).toBe(true);
		expect(timestampWithinSkew(1399, 2000, 300)).toBe(false);
		expect(timestampWithinSkew(2301, 2000, 300)).toBe(false);
	});
});

describe("b64url", () => {
	it("b64url-no-pad 固定向量（Rust 同）", () => {
		expect(b64urlEncode(utf8Bytes("hello"))).toBe("aGVsbG8");
		expect(b64urlEncode(utf8Bytes(""))).toBe("");
		expect(b64urlEncode(new Uint8Array([0xfb, 0xff, 0xfe]))).toBe("-__-");
		expect(b64urlEncode(new Uint8Array([0xfb, 0xff, 0xfe, 0x00]))).toBe("-__-AA");
	});

	it("解码往返一致；非法输入拒绝", () => {
		expect(b64urlEncode(b64urlDecode("aGVsbG8")!)).toBe("aGVsbG8");
		expect(b64urlDecode("aGVsbG8=")).toBeNull(); // 带 padding 拒绝
		expect(b64urlDecode("aGVs+bG8")).toBeNull(); // 标准 base64 字符拒绝
		expect(b64urlDecode("a")).toBeNull(); // 长度非法
	});
});

describe("Crockford 归一化", () => {
	it("示例码 7M4K-P9XQ 归一化为 8 字符", () => {
		expect(normalizeSyncCode("7M4K-P9XQ")).toBe("7M4KP9XQ");
		expect(normalizeSyncCode("7m4k-p9xq")).toBe("7M4KP9XQ");
	});

	it("O→0、I/L→1、大小写不敏感、去连字符；内部空格拒绝（与 Rust 一致）", () => {
		expect(normalizeSyncCode("OOOOOOOO")).toBe("00000000");
		expect(normalizeSyncCode("IIIIIIII")).toBe("11111111");
		expect(normalizeSyncCode("LLLLLLLL")).toBe("11111111");
		expect(normalizeSyncCode(" 7M4KP9XQ ")).toBe("7M4KP9XQ"); // 首尾空白 trim（Rust 同）
		expect(normalizeSyncCode("7M4K P9XQ")).toBeNull(); // 内部空格拒绝（Rust sync_code.rs 同）
	});

	it("拒绝 U、非法字符与长度≠8", () => {
		expect(normalizeSyncCode("7M4KU9XQ")).toBeNull(); // U
		expect(normalizeSyncCode("7M4K-P9X")).toBeNull(); // 长度 7
		expect(normalizeSyncCode("7M4K-P9XQA")).toBeNull(); // 长度 9
		expect(normalizeSyncCode("7M4K*P9XQ")).toBeNull();
		expect(normalizeSyncCode("")).toBeNull();
	});

	it("归一化结果均在字母表内", () => {
		for (const code of ["7M4KP9XQ", "01234567", "ABCDEFGH", "JKMNPQRV", "TWXYZ000"]) {
			const n = normalizeSyncCode(code);
			expect(n).toBe(code);
			expect(normalizeSyncCode(code.toLowerCase())).toBe(code);
		}
	});
});

File: peri-tui/src/sync/canonical_test.rs
#[cfg(test)]
mod tests {
    use crate::sync::canonical::{b64url_nopad, context, timestamp_within_skew, transcript};

    #[test]
    fn test_transcript_fixed_vector() {
        // 固定向量：字段顺序、分隔符与时间戳必须逐字节稳定（TS Worker 复刻）。
        assert_eq!(
            transcript("create", &["ch_abc", "dev_xyz"], 1715000000).unwrap(),
            "peri-sync/v1|create|ch_abc|dev_xyz|1715000000"
        );
        assert_eq!(
            transcript("join", &["code_1", "ch_abc"], 0).unwrap(),
            "peri-sync/v1|join|code_1|ch_abc|0"
        );
    }

    #[test]
    fn test_transcript_empty_fields() {
        assert_eq!(
            transcript("revoke", &[], 42).unwrap(),
            "peri-sync/v1|revoke|42"
        );
    }

    #[test]
    fn test_transcript_rejects_malformed_input() {
        // 入站校验以错误返回而非 panic（Low-4）：空 op、空字段、含 `|`
        // 的 op/字段/尾部一律拒绝。
        assert!(transcript("", &["a"], 1).is_err(), "空 op 必须拒绝");
        assert!(
            transcript("op|evil", &["a"], 1).is_err(),
            "op 含 `|` 必须拒绝"
        );
        assert!(transcript("op", &[""], 1).is_err(), "空字段必须拒绝");
        assert!(transcript("op", &["a", ""], 1).is_err(), "空字段必须拒绝");
        assert!(
            transcript("op", &["a|b"], 1).is_err(),
            "字段含 `|` 必须拒绝"
        );
        assert!(context("op", &["a", ""]).is_err(), "context 空字段必须拒绝");
        assert!(context("op", &[]).is_ok(), "无字段 context 合法");
        assert!(context("op", &["a", "b"]).is_ok(), "合法输入不受影响");
    }

    #[test]
    fn test_context_fixed_vector() {
        assert_eq!(
            context("payload", &["c1", "3", "HASH"]).unwrap(),
            "peri-sync/v1|payload|c1|3|HASH"
        );
        assert_eq!(
            context("keystore", &["salt"]).unwrap(),
            "peri-sync/v1|keystore|salt"
        );
    }

    #[test]
    fn test_transcript_tamper_detection() {
        let base = transcript("create", &["ch_abc", "dev_xyz"], 1715000000).unwrap();
        // 任意字段/操作/时间戳变化都必须改变 transcript。
        assert_ne!(
            base,
            transcript("create", &["ch_abz", "dev_xyz"], 1715000000).unwrap()
        );
        assert_ne!(
            base,
            transcript("create", &["ch_abc", "dev_xyy"], 1715000000).unwrap()
        );
        assert_ne!(
            base,
            transcript("join", &["ch_abc", "dev_xyz"], 1715000000).unwrap()
        );
        assert_ne!(
            base,
            transcript("create", &["ch_abc", "dev_xyz"], 1715000001).unwrap()
        );
    }

    #[test]
    fn test_b64url_nopad_fixed_vector() {
        assert_eq!(b64url_nopad(b"hello"), "aGVsbG8");
        assert_eq!(b64url_nopad(b""), "");
        assert_eq!(b64url_nopad(&[0xFB, 0xFF, 0xFE]), "-__-"); // URL-safe 字符集，无 padding
        assert_eq!(b64url_nopad(&[0xFB, 0xFF, 0xFE, 0x00]), "-__-AA");
    }

    #[test]
    fn test_timestamp_within_skew_boundaries() {
        // 恰好 ±300 秒边界内为真。
        assert!(timestamp_within_skew(1000, 1300, 300));
        assert!(timestamp_within_skew(1000, 700, 300));
        assert!(timestamp_within_skew(1000, 1000, 300));
        // 超过边界为假。
        assert!(!timestamp_within_skew(1000, 1301, 300));
        assert!(!timestamp_within_skew(1000, 699, 300));
        // u64 下溢安全。
        assert!(timestamp_within_skew(0, 0, 0));
        assert!(!timestamp_within_skew(0, 1, 0));
    }
}
